File: src/node/SesSendNotificationSQS.js
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: MIT-0

'use strict'
// Create service client module using ES6 syntax.


import {dynamoUpdateItem, sendEmailFromManifest} from "./utils/index.js";

export const handler = async (event, context) => {
    const requestId = context.awsRequestId
    try {
        const record = event.Records[0]
        const sqsBody = JSON.parse(record.body)
        const {
            id,
            destinatarios,
            manifiesto,
            ConfigurationSetName = "default"
        } = sqsBody
        let responseSes = await sendEmailFromManifest({
            destinatario:destinatarios,
            manifiesto,
            ConfigurationSetName
        });
        const messageId = responseSes.MessageId;
        await dynamoUpdateItem({
            TableName: process.env.TABLE_EMAIL_NAME,
            Key: {
                id: id
            },
            UpdateExpression: "set messageId = :messageId",
            ExpressionAttributeValues: {
                ":messageId": messageId,
            }
        })
        return {
            statusCode: 200
        };
    } catch (err) {
        err.requestId = requestId;
        console.warn(event)
        console.log("Error in writing data to the DynamoDB table : ", err.message)
        throw err
    }
}

File: src/python/src/ses_send_notification_sqs.py
import json
import os

from utils.dynamo import get_dynamo_client
from utils.logic import value_or_default
from utils.s3 import s3_get_object_string
from utils.ses_client import SesClient

table_email_name = os.environ.get('TABLE_EMAIL_NAME')

dynamo_client = get_dynamo_client()

table = dynamo_client.Table(table_email_name)


def handler(message, context):
    try:
        sqsBody = json.loads(message['Records'][0]['body'])

        id = value_or_default(sqsBody, 'id')
        destinatarios = value_or_default(sqsBody, 'destinatarios')
        manifiesto = value_or_default(sqsBody, 'manifiesto')
        ConfigurationSetName = value_or_default(sqsBody, 'ConfigurationSetName', 'default')
        respuesta_email = sen_notification_from_manifest(
            destinatarios,
            manifiesto,
            ConfigurationSetName
        )
        messageId = respuesta_email['MessageId']
        print("MessageId: " + messageId)
        params = {
            'id': id,
        }
        response = table.update_item(
            Key=params,
            UpdateExpression="set messageId = :messageId",
            ExpressionAttributeValues={
                ':messageId': messageId
            },
            ReturnValues="UPDATED_NEW"
        )
        return {
            'statusCode': 200,
            'headers': {}
        }
    except Exception as e:
        print(message)
        print(e)
        raise e


def sen_notification_from_manifest(
        destinatario,
        manifiesto,
        ConfigurationSetName="default"
):
    destinatarios = []
    if isinstance(destinatario, list):
        destinatarios = destinatario
    else:
        destinatarios = destinatario.split(',')
    contenido = s3_get_object_string(manifiesto)[0]
    configuracion_manifiesto = json.loads(contenido)
    emailField = pasar_campos_en_manifiesto_a_objeto(configuracion_manifiesto)

    client = SesClient()
    return client.send_email(
        to_addresses=destinatarios,
        sender_email=emailField['from']['value'],
        subject=emailField['subject']['value'],
        body_text=value_or_default(value_or_default(emailField, 'text', {}), 'value', '-'),
        body_html=emailField['html']['value'],
        attachments=[]
    )


def pasar_campos_en_manifiesto_a_objeto(manifiesto):
    output = {}
    for field in manifiesto['campos']:
        output[field['key']] = field
    return output